File: script/check_sdlc.py
#!/usr/bin/env python3
"""Validate CodeTwo's canonical SDLC artifacts with only the Python standard library."""

from __future__ import annotations

import argparse
import datetime as dt
import re
import subprocess
import sys
from dataclasses import dataclass
from pathlib import Path


REQUIRED_FILES = (
    "docs/sdlc/workflow.md",
    "docs/sdlc/templates/change.md",
    "docs/sdlc/templates/incident.md",
    "docs/sdlc/templates/eval.md",
)

LEGACY_PATHS = (
    "docs/superpowers",
    "docs/sdlc/specs",
    "docs/sdlc/plans",
)

ARTIFACT_KINDS = {
    "changes": "change",
    "incidents": "incident",
    "evals": "eval",
}

ALLOWED_STATUSES = {
    "change": {
        "draft",
        "in-review",
        "accepted",
        "executing",
        "blocked",
        "failed",
        "verified",
        "ready-to-release",
        "released",
        "closed",
        "superseded",
    },
    "incident": {
        "investigating",
        "mitigated",
        "blocked",
        "resolved",
        "closed",
        "superseded",
    },
    "eval": {"draft", "active", "blocked", "retired"},
}

REQUIRED_SECTIONS = {
    "change": (
        "intent",
        "spec",
        "decision and gates",
        "plan",
        "build",
        "verification",
        "review and release",
        "feedback",
    ),
    "incident": (
        "detection and impact",
        "timeline",
        "diagnosis",
        "mitigation and recovery",
        "follow-ups",
        "regression eval",
    ),
    "eval": (
        "provenance",
        "fixed input and environment",
        "allowed actions",
        "observable acceptance",
        "scoring and failure classes",
        "last result",
    ),
}

REQUIRED_FIELDS = (
    "id",
    "kind",
    "status",
    "owner",
    "created",
    "updated",
    "next_trigger",
)

ID_RE = re.compile(r"^[a-z0-9][a-z0-9-]{5,95}$")
HEADING_RE = re.compile(r"^##\s+(.+?)\s*$")
LINK_RE = re.compile(r"\[[^\]]+\]\(([^)]+)\)")
UNASSIGNED = {"", "unassigned", "tbd", "todo", "<owner>"}


@dataclass(frozen=True)
class Artifact:
    path: Path
    metadata: dict[str, str]
    sections: dict[str, str]


def _normalize_heading(value: str) -> str:
    return re.sub(r"\s+", " ", value.strip().lower())


def _unquote(value: str) -> str:
    if len(value) >= 2 and value[0] == value[-1] and value[0] in {'"', "'"}:
        return value[1:-1]
    return value


def parse_artifact(path: Path) -> tuple[Artifact | None, list[str]]:
    errors: list[str] = []
    text = path.read_text(encoding="utf-8")
    lines = text.splitlines()
    relative = path.as_posix()

    if not lines or lines[0].strip() != "---":
        return None, [f"{relative}: missing opening YAML frontmatter delimiter"]

    try:
        closing = next(index for index, line in enumerate(lines[1:], 1) if line.strip() == "---")
    except StopIteration:
        return None, [f"{relative}: missing closing YAML frontmatter delimiter"]

    metadata: dict[str, str] = {}
    for number, line in enumerate(lines[1:closing], 2):
        if not line.strip() or line.lstrip().startswith("#"):
            continue
        match = re.fullmatch(r"([a-z_]+):\s*(.*)", line)
        if not match:
            errors.append(
                f"{relative}:{number}: frontmatter must use flat scalar key: value fields"
            )
            continue
        key, raw_value = match.groups()
        if key in metadata:
            errors.append(f"{relative}:{number}: duplicate frontmatter field {key}")
            continue
        metadata[key] = _unquote(raw_value.strip())

    body_lines = lines[closing + 1 :]
    sections: dict[str, str] = {}
    current: str | None = None
    buffer: list[str] = []

    def flush() -> None:
        nonlocal buffer
        if current is not None:
            sections[current] = "\n".join(buffer).strip()
        buffer = []

    for line in body_lines:
        match = HEADING_RE.match(line)
        if match:
            flush()
            current = _normalize_heading(match.group(1))
        elif current is not None:
            buffer.append(line)
    flush()

    return Artifact(path=path, metadata=metadata, sections=sections), errors


def _validate_local_links(root: Path, path: Path) -> list[str]:
    errors: list[str] = []
    text = path.read_text(encoding="utf-8")
    for target in LINK_RE.findall(text):
        clean = target.strip().strip("<>")
        if not clean or clean.startswith(("#", "https://", "http://", "mailto:")):
            continue
        clean = clean.split("#", 1)[0]
        resolved = (path.parent / clean).resolve()
        try:
            resolved.relative_to(root.resolve())
        except ValueError:
            errors.append(f"{path.as_posix()}: local link escapes repository: {target}")
            continue
        if not resolved.exists():
            errors.append(f"{path.as_posix()}: broken local link: {target}")
    return errors


def validate_artifact(root: Path, artifact: Artifact, expected_kind: str) -> list[str]:
    errors: list[str] = []
    relative = artifact.path.relative_to(root).as_posix()
    metadata = artifact.metadata

    for field in REQUIRED_FIELDS:
        if not metadata.get(field, "").strip():
            errors.append(f"{relative}: missing required frontmatter field {field}")

    kind = metadata.get("kind")
    if kind and kind != expected_kind:
        errors.append(f"{relative}: kind must be {expected_kind}, found {kind}")

    artifact_id = metadata.get("id", "")
    if artifact_id and not ID_RE.fullmatch(artifact_id):
        errors.append(f"{relative}: invalid id {artifact_id!r}")
    expected_id = f"{expected_kind}-{artifact.path.stem}"
    if artifact_id and artifact_id != expected_id:
        errors.append(f"{relative}: id must match filename: expected {expected_id}")

    status = metadata.get("status", "")
    if status and status not in ALLOWED_STATUSES[expected_kind]:
        allowed = ", ".join(sorted(ALLOWED_STATUSES[expected_kind]))
        errors.append(f"{relative}: invalid {expected_kind} status {status!r}; allowed: {allowed}")

    for field in ("created", "updated"):
        value = metadata.get(field, "")
        if value:
            try:
                dt.date.fromisoformat(value)
            except ValueError:
                errors.append(f"{relative}: {field} must be YYYY-MM-DD, found {value!r}")

    if metadata.get("created") and metadata.get("updated"):
        if metadata["updated"] < metadata["created"]:
            errors.append(f"{relative}: updated date precedes created date")

    if status not in {"", "draft", "in-review", "investigating"}:
        if metadata.get("owner", "").strip().lower() in UNASSIGNED:
            errors.append(f"{relative}: status {status} requires an assigned owner")

    for heading in REQUIRED_SECTIONS[expected_kind]:
        if heading not in artifact.sections:
            errors.append(f"{relative}: missing required section ## {heading.title()}")
        elif not artifact.sections[heading].strip():
            errors.append(f"{relative}: section ## {heading.title()} is empty")

    if expected_kind == "change" and status in {
        "verified",
        "ready-to-release",
        "released",
        "closed",
    }:
        verification = artifact.sections.get("verification", "").lower()
        if re.search(r"(?m)^\s*(pending|todo)(?:\s|$)", verification):
            errors.append(f"{relative}: status {status} cannot have pending verification")

    errors.extend(_validate_local_links(root, artifact.path))
    return errors


def _changed_paths(root: Path, base: str) -> tuple[list[tuple[str, tuple[str, ...]]], list[str]]:
    result = subprocess.run(
        ["git", "diff", "--name-status", "--find-renames", f"{base}...HEAD"],
        cwd=root,
        text=True,
        capture_output=True,
        check=False,
    )
    if result.returncode != 0:
        detail = result.stderr.strip() or result.stdout.strip() or "git diff failed"
        return [], [f"cannot compare SDLC branch with base {base}: {detail}"]

    changes: list[tuple[str, tuple[str, ...]]] = []
    for line in result.stdout.splitlines():
        parts = line.split("\t")
        if len(parts) < 2:
            continue
        changes.append((parts[0], tuple(parts[1:])))
    return changes, []


def _validate_changed_artifact_gate(root: Path, base: str) -> list[str]:
    changes, errors = _changed_paths(root, base)
    if errors or not changes:
        return errors

    material_paths = {path for _, paths in changes for path in paths}
    changed_change_artifacts = {
        path
        for status, paths in changes
        if not status.startswith("D")
        for path in paths
        if path.startswith("docs/sdlc/changes/") and path.endswith(".md")
    }

    if material_paths and not changed_change_artifacts:
        return [
            "material branch changes require an added or updated canonical file under "
            "docs/sdlc/changes/"
        ]
    return []


def _validate_release_gate(artifacts: dict[str, Artifact], change_id: str) -> list[str]:
    artifact = artifacts.get(change_id)
    if artifact is None:
        return [f"release change artifact not found: {change_id}"]
    if artifact.metadata.get("kind") != "change":
        return [f"release artifact must be a change: {change_id}"]
    status = artifact.metadata.get("status")
    if status != "ready-to-release":
        return [
            f"release change {change_id} must be ready-to-release, found {status or 'missing'}"
        ]
    return []


def validate_repository(
    root: Path,
    base: str | None = None,
    release_change: str | None = None,
) -> list[str]:
    root = root.resolve()
    errors: list[str] = []

    for relative in REQUIRED_FILES:
        if not (root / relative).is_file():
            errors.append(f"missing required SDLC file: {relative}")

    for relative in LEGACY_PATHS:
        if (root / relative).exists():
            errors.append(f"legacy or parallel lifecycle path is forbidden: {relative}")

    seen_ids: dict[str, str] = {}
    artifacts: dict[str, Artifact] = {}
    change_count = 0
    for directory, expected_kind in ARTIFACT_KINDS.items():
        artifact_root = root / "docs" / "sdlc" / directory
        if not artifact_root.exists():
            continue
        for path in sorted(artifact_root.rglob("*.md")):
            artifact, parse_errors = parse_artifact(path)
            errors.extend(parse_errors)
            if artifact is None:
                continue
            errors.extend(validate_artifact(root, artifact, expected_kind))
            artifact_id = artifact.metadata.get("id")
            relative = path.relative_to(root).as_posix()
            if artifact_id:
                if artifact_id in seen_ids:
                    errors.append(
                        f"duplicate artifact id {artifact_id!r}: {seen_ids[artifact_id]} and {relative}"
                    )
                else:
                    seen_ids[artifact_id] = relative
                    artifacts[artifact_id] = artifact
            if expected_kind == "change":
                change_count += 1

    if change_count == 0:
        errors.append("at least one canonical change artifact is required")

    workflow = root / "docs" / "sdlc" / "workflow.md"
    if workflow.is_file():
        errors.extend(_validate_local_links(root, workflow))

    if base:
        errors.extend(_validate_changed_artifact_gate(root, base))
    if release_change:
        errors.extend(_validate_release_gate(artifacts, release_change))

    return errors


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--root",
        type=Path,
        default=Path(__file__).resolve().parents[1],
        help="repository root (defaults to the parent of script/)",
    )
    parser.add_argument(
        "--base",
        help="optional Git base commit; require a changed canonical artifact for branch changes",
    )
    parser.add_argument(
        "--release-change",
        help="require this change Artifact to be in ready-to-release state",
    )
    args = parser.parse_args(argv)

    errors = validate_repository(args.root, args.base, args.release_change)
    if errors:
        for error in errors:
            print(f"[sdlc] error: {error}", file=sys.stderr)
        print(f"[sdlc] failed with {len(errors)} error(s)", file=sys.stderr)
        return 1

    print("[sdlc] contract valid")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: script/check-sdlc.ts
#!/usr/bin/env bun

import { spawnSync } from "node:child_process";
import { existsSync, readFileSync, readdirSync, statSync } from "node:fs";
import { dirname, isAbsolute, join, relative, resolve, sep } from "node:path";
import { fileURLToPath } from "node:url";

export const REQUIRED_FILES = [
  "docs/sdlc/workflow.md",
  "docs/sdlc/templates/change.md",
  "docs/sdlc/templates/incident.md",
  "docs/sdlc/templates/eval.md",
] as const;

const LEGACY_PATHS = [
  "docs/superpowers",
  "docs/sdlc/specs",
  "docs/sdlc/plans",
  "docs/sdlc/changes/2026-08-29-sdlc-bootstrap.md",
  "docs/sdlc/evals/legacy-workflow-single-source.md",
] as const;

const ARTIFACT_KINDS = {
  changes: "change",
  incidents: "incident",
  evals: "eval",
} as const;

type ArtifactKind = (typeof ARTIFACT_KINDS)[keyof typeof ARTIFACT_KINDS];

const ALLOWED_STATUSES: Record<ArtifactKind, Set<string>> = {
  change: new Set([
    "draft",
    "in-review",
    "accepted",
    "executing",
    "blocked",
    "failed",
    "verified",
    "ready-to-release",
    "released",
    "closed",
    "superseded",
  ]),
  incident: new Set([
    "draft",
    "investigating",
    "mitigated",
    "blocked",
    "resolved",
    "closed",
    "superseded",
  ]),
  eval: new Set(["draft", "active", "blocked", "failed", "retired"]),
};

const REQUIRED_SECTIONS: Record<ArtifactKind, readonly string[]> = {
  change: [
    "intent",
    "spec",
    "decision and gates",
    "plan",
    "build",
    "verification",
    "review and release",
    "feedback",
  ],
  incident: [
    "detection and impact",
    "timeline",
    "diagnosis",
    "mitigation and recovery",
    "follow-ups",
    "regression eval",
  ],
  eval: [
    "provenance",
    "fixed input and environment",
    "allowed actions",
    "observable acceptance",
    "scoring and failure classes",
    "last result",
  ],
};

const REQUIRED_FIELDS = [
  "id",
  "kind",
  "status",
  "owner",
  "approvers",
  "created",
  "updated",
  "source",
  "inputs",
  "outputs",
  "next_trigger",
] as const;

const ADVANCED_CHANGE_STATUSES = new Set([
  "accepted",
  "executing",
  "failed",
  "verified",
  "ready-to-release",
  "released",
  "closed",
]);
const BUILD_CHANGE_STATUSES = new Set([
  "executing",
  "failed",
  "verified",
  "ready-to-release",
  "released",
  "closed",
]);
const VERIFIED_CHANGE_STATUSES = new Set([
  "verified",
  "ready-to-release",
  "released",
  "closed",
]);

const ID_RE = /^[a-z0-9][a-z0-9-]{5,95}$/;
const TITLE_RE = /^#\s+(.+?)\s*$/;
const HEADING_RE = /^##\s+(.+?)\s*$/;
const CHECKBOX_RE = /^\s*-\s+\[([ xX])\]\s+\S/gm;
const PLACEHOLDER_RE = /^(?:pending|todo|tbd|unassigned|none recorded|not applicable|n\/a|<[^>]+>)(?:[.:\s].*)?$/i;
const UNASSIGNED = new Set([
  "",
  "[]",
  "none",
  "none recorded",
  "unassigned",
  "tbd",
  "todo",
  "<owner>",
]);

interface Artifact {
  path: string;
  title: string;
  metadata: Record<string, string>;
  sections: Record<string, string>;
}

interface ParsedArtifact {
  artifact: Artifact | null;
  errors: string[];
}

function normalizeHeading(value: string): string {
  return value.trim().toLowerCase().replace(/\s+/g, " ");
}

function unquote(value: string): string {
  if (
    value.length >= 2 &&
    value[0] === value.at(-1) &&
    (value[0] === '"' || value[0] === "'")
  ) {
    return value.slice(1, -1);
  }
  return value;
}

function repoPath(root: string, path: string): string {
  return relative(root, path).split(sep).join("/");
}

function parseArtifact(path: string): ParsedArtifact {
  const errors: string[] = [];
  const lines = readFileSync(path, "utf8").split(/\r?\n/);
  const displayPath = path.split(sep).join("/");

  if (lines[0]?.trim() !== "---") {
    return {
      artifact: null,
      errors: [`${displayPath}: missing opening YAML frontmatter delimiter`],
    };
  }

  const closing = lines.slice(1).findIndex((line) => line.trim() === "---") + 1;
  if (closing === 0) {
    return {
      artifact: null,
      errors: [`${displayPath}: missing closing YAML frontmatter delimiter`],
    };
  }

  const metadata: Record<string, string> = {};
  for (let index = 1; index < closing; index += 1) {
    const line = lines[index] ?? "";
    if (!line.trim() || line.trimStart().startsWith("#")) continue;
    const match = /^([a-z_]+):\s*(.*)$/.exec(line);
    if (!match) {
      errors.push(
        `${displayPath}:${index + 1}: frontmatter must use flat scalar key: value fields`,
      );
      continue;
    }
    const [, key, rawValue] = match;
    if (key in metadata) {
      errors.push(`${displayPath}:${index + 1}: duplicate frontmatter field ${key}`);
      continue;
    }
    metadata[key] = unquote(rawValue.trim());
  }

  let title = "";
  let current: string | null = null;
  let buffer: string[] = [];
  const sections: Record<string, string> = {};
  const flush = () => {
    if (current !== null) {
      if (current in sections) {
        errors.push(`${displayPath}: duplicate section ## ${titleCase(current)}`);
      } else {
        sections[current] = buffer.join("\n").trim();
      }
    }
    buffer = [];
  };

  for (const line of lines.slice(closing + 1)) {
    const titleMatch = TITLE_RE.exec(line);
    const headingMatch = HEADING_RE.exec(line);
    if (titleMatch && !title) {
      title = titleMatch[1].trim();
    } else if (headingMatch) {
      flush();
      current = normalizeHeading(headingMatch[1]);
    } else if (current !== null) {
      buffer.push(line);
    }
  }
  flush();

  return { artifact: { path, title, metadata, sections }, errors };
}

function titleCase(value: string): string {
  return value.replace(/\b\w/g, (letter) => letter.toUpperCase());
}

function linkTargets(text: string): string[] {
  return Array.from(text.matchAll(/\[[^\]]+\]\(([^)]+)\)/g), (match) => match[1]);
}

function validateLocalLinks(root: string, path: string): string[] {
  const errors: string[] = [];
  const displayPath = path.split(sep).join("/");
  for (const target of linkTargets(readFileSync(path, "utf8"))) {
    let clean = target.trim().replace(/^<|>$/g, "");
    if (!clean || /^(?:#|https?:\/\/|mailto:)/.test(clean)) continue;
    clean = clean.split("#", 1)[0];
    const resolved = resolve(dirname(path), clean);
    const fromRoot = relative(root, resolved);
    if (fromRoot === ".." || fromRoot.startsWith(`..${sep}`) || isAbsolute(fromRoot)) {
      errors.push(`${displayPath}: local link escapes repository: ${target}`);
    } else if (!existsSync(resolved)) {
      errors.push(`${displayPath}: broken local link: ${target}`);
    }
  }
  return errors;
}

function labelValue(section: string, label: string): string | null {
  const escaped = label.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  return new RegExp(`^\\s*${escaped}:\\s*(.+?)\\s*$`, "im").exec(section)?.[1].trim() ?? null;
}

function isConcrete(value: string | null | undefined): boolean {
  return Boolean(value && !PLACEHOLDER_RE.test(value.trim()));
}

function hasLinkTo(section: string, pathPart: string): boolean {
  return linkTargets(section).some((target) => target.includes(pathPart));
}

function hasBlocker(section: string): boolean {
  return isConcrete(labelValue(section, "Blocked"));
}

function hasVerificationEvidence(section: string): boolean {
  const hasResult = /\b(?:pass(?:ed)?|fail(?:ed)?|blocked|verified)\b/i.test(section);
  const hasSubject = section.includes("`") || linkTargets(section).length > 0 || section.includes("|");
  return hasResult && hasSubject;
}

function validateChange(artifact: Artifact, path: string): string[] {
  const errors: string[] = [];
  const status = artifact.metadata.status ?? "";
  const spec = artifact.sections.spec ?? "";
  const decision = artifact.sections["decision and gates"] ?? "";
  const verification = artifact.sections.verification ?? "";
  const review = artifact.sections["review and release"] ?? "";
  const criteria = Array.from(spec.matchAll(CHECKBOX_RE), (match) => match[1]);

  if (criteria.length === 0) {
    errors.push(`${path}: change Spec must contain observable acceptance checkboxes`);
  }

  if (ADVANCED_CHANGE_STATUSES.has(status)) {
    if (UNASSIGNED.has((artifact.metadata.approvers ?? "").trim().toLowerCase())) {
      errors.push(`${path}: status ${status} requires named Intent/Spec approval`);
    }
    for (const field of ["source", "inputs", "outputs", "next_trigger"]) {
      if (!isConcrete(artifact.metadata[field])) {
        errors.push(`${path}: status ${status} requires concrete ${field} metadata`);
      }
    }
    if (!decision.trim() || /^\s*(?:pending|todo|tbd)\.?\s*$/i.test(decision)) {
      errors.push(`${path}: status ${status} requires Decision and Gate evidence`);
    }
  }

  if (VERIFIED_CHANGE_STATUSES.has(status)) {
    if (criteria.some((mark) => mark.trim().toLowerCase() !== "x")) {
      errors.push(`${path}: status ${status} requires every acceptance criterion checked`);
    }
    if ((labelValue(verification, "Verdict") ?? "").toLowerCase().replace(/\.$/, "") !== "verified") {
      errors.push(`${path}: status ${status} requires 'Verdict: verified'`);
    }
    if (!isConcrete(labelValue(verification, "Residual risk"))) {
      errors.push(`${path}: status ${status} requires a concrete Residual risk`);
    }
    if (!hasVerificationEvidence(verification)) {
      errors.push(`${path}: status ${status} requires actual verification evidence`);
    }
  }

  if (status === "failed") {
    if ((labelValue(verification, "Verdict") ?? "").toLowerCase().replace(/\.$/, "") !== "failed") {
      errors.push(`${path}: failed status requires 'Verdict: failed'`);
    }
    if (!hasVerificationEvidence(verification)) {
      errors.push(`${path}: failed status requires the observed failed evidence`);
    }
  }

  if (status === "ready-to-release" || status === "released") {
    if (!isConcrete(labelValue(review, "Approval"))) {
      errors.push(`${path}: status ${status} requires explicit release Approval`);
    }
    const target = labelValue(review, "Release target");
    if (!isConcrete(target) || target?.toLowerCase().replace(/\.$/, "") === "none") {
      errors.push(`${path}: status ${status} requires a concrete Release target`);
    }
    if (!isConcrete(labelValue(review, "Rollback"))) {
      errors.push(`${path}: status ${status} requires a concrete Rollback path`);
    }
  }

  if (status === "released") {
    if (!isConcrete(labelValue(review, "Release identity"))) {
      errors.push(`${path}: released status requires an immutable Release identity`);
    }
    if (!isConcrete(labelValue(review, "Smoke evidence"))) {
      errors.push(`${path}: released status requires observed Smoke evidence`);
    }
  }

  if (status === "closed") {
    const hasRelease =
      isConcrete(labelValue(review, "Release identity")) &&
      isConcrete(labelValue(review, "Smoke evidence"));
    if (!hasRelease && !isConcrete(labelValue(review, "No release"))) {
      errors.push(
        `${path}: closed status requires release evidence or a concrete No release disposition`,
      );
    }
  }

  return errors;
}

function validateIncident(artifact: Artifact, path: string): string[] {
  const errors: string[] = [];
  const status = artifact.metadata.status ?? "";
  if (status !== "resolved" && status !== "closed") return errors;

  for (const field of ["source", "inputs", "outputs", "next_trigger"]) {
    if (!isConcrete(artifact.metadata[field])) {
      errors.push(`${path}: status ${status} requires concrete ${field} metadata`);
    }
  }

  const recovery = artifact.sections["mitigation and recovery"] ?? "";
  const followUps = artifact.sections["follow-ups"] ?? "";
  const regression = artifact.sections["regression eval"] ?? "";
  if ((labelValue(recovery, "Recovery verdict") ?? "").toLowerCase().replace(/\.$/, "") !== "recovered") {
    errors.push(`${path}: status ${status} requires 'Recovery verdict: recovered'`);
  }
  if (!hasLinkTo(followUps, "/changes/") && !hasBlocker(followUps)) {
    errors.push(`${path}: status ${status} requires a linked follow-up change or Blocked reason`);
  }
  if (!hasLinkTo(regression, "/evals/") && !hasBlocker(regression)) {
    errors.push(`${path}: status ${status} requires a linked regression Eval or Blocked reason`);
  }
  return errors;
}

function validateEval(artifact: Artifact, path: string): string[] {
  const errors: string[] = [];
  const status = artifact.metadata.status ?? "";
  if (!new Set(["active", "failed", "retired"]).has(status)) return errors;

  for (const field of ["source", "inputs", "outputs", "next_trigger"]) {
    if (!isConcrete(artifact.metadata[field])) {
      errors.push(`${path}: status ${status} requires concrete ${field} metadata`);
    }
  }

  const provenance = artifact.sections.provenance ?? "";
  const lastResult = artifact.sections["last result"] ?? "";
  if (linkTargets(provenance).length === 0) {
    errors.push(`${path}: status ${status} requires linked real-task or Incident provenance`);
  }
  const result = (labelValue(lastResult, "Result") ?? "").toLowerCase().replace(/\.$/, "");
  if (!new Set(["pass", "fail", "blocked"]).has(result)) {
    errors.push(`${path}: status ${status} requires 'Result: pass', 'fail', or 'blocked'`);
  }
  if (!isConcrete(labelValue(lastResult, "Revision"))) {
    errors.push(`${path}: status ${status} requires the tested Revision`);
  }
  return errors;
}

function isValidDate(value: string): boolean {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(value)) return false;
  const date = new Date(`${value}T00:00:00Z`);
  return !Number.isNaN(date.valueOf()) && date.toISOString().slice(0, 10) === value;
}

function validateArtifact(root: string, artifact: Artifact, expectedKind: ArtifactKind): string[] {
  const errors: string[] = [];
  const path = repoPath(root, artifact.path);
  const metadata = artifact.metadata;

  if (!artifact.title) errors.push(`${path}: missing one level-one Artifact title`);

  for (const field of REQUIRED_FIELDS) {
    if (!(metadata[field] ?? "").trim()) {
      errors.push(`${path}: missing required frontmatter field ${field}`);
    }
  }

  const kind = metadata.kind;
  if (kind && kind !== expectedKind) {
    errors.push(`${path}: kind must be ${expectedKind}, found ${kind}`);
  }

  const id = metadata.id ?? "";
  if (id && !ID_RE.test(id)) errors.push(`${path}: invalid id ${JSON.stringify(id)}`);
  const expectedId = `${expectedKind}-${artifact.path.split(sep).at(-1)?.replace(/\.md$/, "")}`;
  if (id && id !== expectedId) {
    errors.push(`${path}: id must match filename: expected ${expectedId}`);
  }

  const status = metadata.status ?? "";
  if (status && !ALLOWED_STATUSES[expectedKind].has(status)) {
    const allowed = Array.from(ALLOWED_STATUSES[expectedKind]).sort().join(", ");
    errors.push(`${path}: invalid ${expectedKind} status ${JSON.stringify(status)}; allowed: ${allowed}`);
  }

  for (const field of ["created", "updated"]) {
    const value = metadata[field] ?? "";
    if (value && !isValidDate(value)) {
      errors.push(`${path}: ${field} must be YYYY-MM-DD, found ${JSON.stringify(value)}`);
    }
  }
  if (metadata.created && metadata.updated && metadata.updated < metadata.created) {
    errors.push(`${path}: updated date precedes created date`);
  }

  if (!["", "draft", "in-review", "investigating"].includes(status)) {
    if (UNASSIGNED.has((metadata.owner ?? "").trim().toLowerCase())) {
      errors.push(`${path}: status ${status} requires an assigned owner`);
    }
  }

  for (const heading of REQUIRED_SECTIONS[expectedKind]) {
    if (!(heading in artifact.sections)) {
      errors.push(`${path}: missing required section ## ${titleCase(heading)}`);
    } else if (!artifact.sections[heading].trim()) {
      errors.push(`${path}: section ## ${titleCase(heading)} is empty`);
    }
  }

  if (expectedKind === "change") errors.push(...validateChange(artifact, path));
  if (expectedKind === "incident") errors.push(...validateIncident(artifact, path));
  if (expectedKind === "eval") errors.push(...validateEval(artifact, path));
  errors.push(...validateLocalLinks(root, artifact.path));
  return errors;
}

function markdownFiles(root: string): string[] {
  if (!existsSync(root)) return [];
  const files: string[] = [];
  for (const entry of readdirSync(root)) {
    const path = join(root, entry);
    if (statSync(path).isDirectory()) files.push(...markdownFiles(path));
    else if (entry.endsWith(".md")) files.push(path);
  }
  return files.sort();
}

interface ChangedPath {
  status: string;
  paths: string[];
}

function changedPaths(root: string, base: string): { changes: ChangedPath[]; errors: string[] } {
  const result = spawnSync(
    "git",
    ["diff", "--name-status", "--find-renames", `${base}...HEAD`],
    { cwd: root, encoding: "utf8" },
  );
  if (result.status !== 0) {
    const detail = result.stderr.trim() || result.stdout.trim() || "git diff failed";
    return { changes: [], errors: [`cannot compare SDLC branch with base ${base}: ${detail}`] };
  }
  const changes = result.stdout
    .split(/\r?\n/)
    .filter(Boolean)
    .map((line) => {
      const [status, ...paths] = line.split("\t");
      return { status, paths };
    })
    .filter((change) => change.paths.length > 0);
  return { changes, errors: [] };
}

function validateChangedArtifactGate(
  root: string,
  base: string,
  artifactsByPath: Map<string, Artifact>,
): string[] {
  const { changes, errors } = changedPaths(root, base);
  if (errors.length > 0 || changes.length === 0) return errors;

  const changed = new Set(changes.flatMap((change) => change.paths));
  const changedArtifacts = new Set(
    changes
      .filter((change) => !change.status.startsWith("D"))
      .flatMap((change) => change.paths)
      .filter((path) => path.startsWith("docs/sdlc/changes/") && path.endsWith(".md")),
  );
  if (changedArtifacts.size === 0) {
    return ["branch changes require an added or updated canonical file under docs/sdlc/changes/"];
  }

  if (Array.from(changed).every((path) => changedArtifacts.has(path))) return [];
  const ready = Array.from(changedArtifacts).some((path) => {
    const artifact = artifactsByPath.get(path);
    return artifact && BUILD_CHANGE_STATUSES.has(artifact.metadata.status ?? "");
  });
  if (!ready) {
    return [
      "branch implementation changes require a changed canonical Artifact in executing, failed, verified, ready-to-release, released, or closed state",
    ];
  }
  return [];
}

function validateReleaseGate(artifacts: Map<string, Artifact>, changeId: string): string[] {
  const artifact = artifacts.get(changeId);
  if (!artifact) return [`release change artifact not found: ${changeId}`];
  if (artifact.metadata.kind !== "change") {
    return [`release artifact must be a change: ${changeId}`];
  }
  const status = artifact.metadata.status;
  if (status !== "ready-to-release") {
    return [`release change ${changeId} must be ready-to-release, found ${status || "missing"}`];
  }
  return [];
}

export function validateRepository(
  repositoryRoot: string,
  base?: string,
  releaseChange?: string,
): string[] {
  const root = resolve(repositoryRoot);
  const errors: string[] = [];

  for (const path of REQUIRED_FILES) {
    if (!existsSync(join(root, path)) || !statSync(join(root, path)).isFile()) {
      errors.push(`missing required SDLC file: ${path}`);
    }
  }
  for (const path of LEGACY_PATHS) {
    if (existsSync(join(root, path))) {
      errors.push(`legacy or superseded lifecycle path is forbidden: ${path}`);
    }
  }

  const seenIds = new Map<string, string>();
  const artifacts = new Map<string, Artifact>();
  const artifactsByPath = new Map<string, Artifact>();
  let changeCount = 0;

  for (const [directory, expectedKind] of Object.entries(ARTIFACT_KINDS)) {
    for (const path of markdownFiles(join(root, "docs", "sdlc", directory))) {
      const parsed = parseArtifact(path);
      errors.push(...parsed.errors);
      if (!parsed.artifact) continue;
      const artifact = parsed.artifact;
      errors.push(...validateArtifact(root, artifact, expectedKind));
      const id = artifact.metadata.id;
      const relativePath = repoPath(root, path);
      artifactsByPath.set(relativePath, artifact);
      if (id) {
        const previous = seenIds.get(id);
        if (previous) {
          errors.push(`duplicate artifact id ${JSON.stringify(id)}: ${previous} and ${relativePath}`);
        } else {
          seenIds.set(id, relativePath);
          artifacts.set(id, artifact);
        }
      }
      if (expectedKind === "change") changeCount += 1;
    }
  }

  if (changeCount === 0) errors.push("at least one canonical change artifact is required");
  const workflow = join(root, "docs", "sdlc", "workflow.md");
  if (existsSync(workflow)) errors.push(...validateLocalLinks(root, workflow));
  if (base) errors.push(...validateChangedArtifactGate(root, base, artifactsByPath));
  if (releaseChange) errors.push(...validateReleaseGate(artifacts, releaseChange));
  return errors;
}

function parseArguments(argv: string[]): { root: string; base?: string; releaseChange?: string } {
  let root = resolve(dirname(fileURLToPath(import.meta.url)), "..");
  let base: string | undefined;
  let releaseChange: string | undefined;
  for (let index = 0; index < argv.length; index += 1) {
    const argument = argv[index];
    const value = argv[index + 1];
    if (!["--root", "--base", "--release-change"].includes(argument) || !value) {
      throw new Error(`usage: bun script/check-sdlc.ts [--root PATH] [--base SHA] [--release-change ID]`);
    }
    if (argument === "--root") root = value;
    if (argument === "--base") base = value;
    if (argument === "--release-change") releaseChange = value;
    index += 1;
  }
  return { root, base, releaseChange };
}

if (import.meta.main) {
  try {
    const args = parseArguments(process.argv.slice(2));
    const errors = validateRepository(args.root, args.base, args.releaseChange);
    if (errors.length > 0) {
      for (const error of errors) console.error(`[sdlc] error: ${error}`);
      console.error(`[sdlc] failed with ${errors.length} error(s)`);
      process.exit(1);
    }
    console.log("[sdlc] contract valid");
  } catch (error) {
    console.error(`[sdlc] error: ${error instanceof Error ? error.message : String(error)}`);
    process.exit(2);
  }
}
